framework/include/utils/SymmetricRankFourTensor.h

```diff
@@ -0,0 +1,581 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "Moose.h"
+#include "ADRankTwoTensorForward.h"
+#include "ADRankFourTensorForward.h"
+#include "ADSymmetricRankTwoTensorForward.h"
+#include "ADSymmetricRankFourTensorForward.h"
+#include "MooseUtils.h"
+#include "MathUtils.h"
+
+#include "libmesh/libmesh.h"
+#include "libmesh/tuple_of.h"
+#include "libmesh/int_range.h"
+
+#include "metaphysicl/raw_type.h"
+
+#include <petscsys.h>
+
+// Eigen includes
+#include <Eigen/Core>
+#include <Eigen/Dense>
+
+#include <array>
+
+using libMesh::Real;
+using libMesh::tuple_of;
+namespace libMesh
+{
+template <typename>
+class TensorValue;
+template <typename>
+class TypeTensor;
+template <typename>
+class VectorValue;
+}
+
+// Forward declarations
+class MooseEnum;
+
+namespace MathUtils
+{
+/**
+ * Helper function template specialization to set an object to zero.
+ * Needed by DerivativeMaterialInterface
+ */
+template <>
+void mooseSetToZero<SymmetricRankFourTensor>(SymmetricRankFourTensor & v);
+template <>
+void mooseSetToZero<ADSymmetricRankFourTensor>(ADSymmetricRankFourTensor & v);
+}
+
+/**
+ * SymmetricRankFourTensorTempl is designed to handle any N-dimensional fourth order tensor, C.
+ *
+ * It is designed to allow for maximum clarity of the mathematics and ease of use.
+ * Original class authors: A. M. Jokisaari, O. Heinonen, M.R. Tonks
```

Comment:
I have not objection to giving credit where it is due, but don't we track authorship with the git history? I'd say we should delete that since we don't typically add authorship lines in any of our other code. Plus, I think this is just a copy/paste from the other class.